soroban-env-host/src/native_contract/token/balance.rs

```diff
@@ -31,77 +44,216 @@ fn write_balance(e: &Host, id: Identifier, amount: i128) -> Result<(), HostError
+                        ContractError::BalanceError,
+                        "trustline isn't clawbackable",
+                    ));
+                }
+                Ok(())
+            }
+            Metadata::AlphaNum12(asset) => {
+                if asset.issuer == acc_id {
+                    return Err(e.err_status_msg(
+                        ContractError::OperationNotSupportedError,
+                        "cannot clawback from issuer",
+                    ));
+                }
+                let tl_flags = get_trustline_flags(
+                    e,
+                    acc_id,
+                    e.create_asset_12(asset.asset_code.to_array()?, asset.issuer),
+                )?;
+                if tl_flags & (TrustLineFlags::TrustlineClawbackEnabledFlag as u32) == 0 {
+                    return Err(e.err_status_msg(
+                        ContractError::BalanceError,
+                        "trustline isn't clawbackable",
+                    ));
+                }
+                Ok(())
+            }
+        },
+        _ => {
+            // TODO: Non-account balances are always clawbackable for now if admin is set. Revisit this.
+            Ok(())
+        }
+    }
 }
 
 // Metering: covered by components
 pub fn transfer_classic_balance(e: &Host, to_key: AccountId, amount: i64) -> Result<(), HostError> {
     match read_metadata(e)? {
         Metadata::Native => transfer_account_balance(e, to_key, amount)?,
-        Metadata::AlphaNum4(asset) => transfer_trustline_balance(
-            e,
-            to_key,
-            e.create_asset_4(asset.asset_code.to_array()?, asset.issuer),
-            amount,
-        )?,
-        Metadata::AlphaNum12(asset) => transfer_trustline_balance(
-            e,
-            to_key,
-            e.create_asset_12(asset.asset_code.to_array()?, asset.issuer),
-            amount,
-        )?,
+        Metadata::AlphaNum4(asset) => {
+            if asset.issuer == to_key {
+                return Ok(());
+            }
+
+            transfer_trustline_balance(
+                e,
+                to_key,
+                e.create_asset_4(asset.asset_code.to_array()?, asset.issuer),
+                amount,
+            )?
+        }
+        Metadata::AlphaNum12(asset) => {
+            if asset.issuer == to_key {
+                return Ok(());
+            }
+
+            transfer_trustline_balance(
+                e,
+                to_key,
+                e.create_asset_12(asset.asset_code.to_array()?, asset.issuer),
+                amount,
+            )?
+        }
     };
     Ok(())
 }
 
+// TODO: Metering analysis
+//returns (total balance, spendable balance)
+pub fn get_classic_balance(e: &Host, to_key: AccountId) -> Result<(i64, i64), HostError> {
+    match read_metadata(e)? {
+        Metadata::Native => get_account_balance(e, to_key),
+        Metadata::AlphaNum4(asset) => {
+            if asset.issuer == to_key {
```

Comment:
I'm actually not clear on what this helper would look like, and how it would be used here. The asset type is different between the match cases and we're just checking the issuer. Are you saying we should have a helper that returns the `TrustLineAsset` and issuer based on the metadata?

Reply:
Oh wait I see what you're saying. I should use a method that takes the issuer and TrustLineAsset so all special logic is shared.

Reply:
Yeah, exactly, It wouldn't save too many lines, but would make maintenance/review way easier.

Reply:
Done